MediaBrowser.Api/Playback/Hls/DynamicHlsService.cs

```diff
@@ -927,35 +927,41 @@ protected override string GetVideoArguments(StreamState state, EncodingOptions e
             }
             else
             {
+                var gopArg = string.Empty;
                 var keyFrameArg = string.Format(
                     CultureInfo.InvariantCulture,
                     " -force_key_frames:0 \"expr:gte(t,{0}+n_forced*{1})\"",
                     GetStartNumber(state) * state.SegmentLength,
                     state.SegmentLength);
-                if (state.TargetFramerate.HasValue)
+
+                var framerate = (state.VideoStream != null) ? state.VideoStream.RealFrameRate : null;
+
+                if (framerate.HasValue)
                 {
                     // This is to make sure keyframe interval is limited to our segment,
                     // as forcing keyframes is not enough.
                     // Example: we encoded half of desired length, then codec detected
                     // scene cut and inserted a keyframe; next forced keyframe would
                     // be created outside of segment, which breaks seeking.
-                    keyFrameArg += string.Format(
+                    gopArg = string.Format(
                         CultureInfo.InvariantCulture,
-                        " -g {0} -keyint_min {0}",
-                        (int)(state.SegmentLength * state.TargetFramerate)
+                        " -g {0} -keyint_min {0} -sc_threshold 0",
+                        state.SegmentLength * Math.Ceiling(Convert.ToDecimal(framerate))
```

Comment:
Can you please explain both changes? What does `sc_threshold` do, and why did you have to do a `Math.Ceiling` instead of just casting to an int?

Reply:
```
hls slice duration = Min(hls_time, framerate/gop_size)

hls_time <= framerate/gop_size

gop_size <= framerate/hls_time 
```

If this condition is not met, the hls slice duration is not the usual 3, but the wrong 10. And it makes the size of each slice larger, which is not conducive to fast loading. You can see it in the `.m3u8` file in transcodes folder.

For these hardware encoders, `force_keyframe` is invalid, so we need to manually set gop (keyframe) and `key minimum`. 

`sc_threshold 0` is used to prevent the hardware encoder from post processing to break the set keyframe.

Casting will lose precision and make `hls_time` greater than another value, and will also result in incorrect slicing duration. For example, for a video with a frame rate of 23.976, set `hls_time` to 3, which will result in an incorrect `gop_size` of 71 and a slice duration of 6. The correct `gop_size` should be 72 (3 * 24).

Reply:
Got it on `sc_threshold`, would be good to have it as a code comment.

I don't get why it would produce slice duration of 6 in above example. It doesn't make sense to me to multiply framerate and gop as you did (because framerate is `frames/sec` and `gop` is `frames`, so you end up with `1/sec` instead of `sec` - wrong dimension; you should be doing `gop/framerate`). So in your example above I think we'd end with 2.96 seconds hls segment, not a big difference.
Also it's unclear why you want to convert to `Decimal` and then ceil, why not ceil immediately?

All in all, maybe change it to `Math.Ceiling(state.SegmentLength * framerate)` - it looks cleaner and easier to understand, and would produce a more correct result anyway.

Reply:
> I don't get why it would produce slice duration of 6 in above example. It doesn't make sense to me to multiply framerate and gop as you did (because framerate is `frames/sec` and `gop` is `frames`, so you end up with `1/sec` instead of `sec` - wrong dimension; you should be doing `gop/framerate`). So in your example above I think we'd end with 2.96 seconds hls segment, not a big difference.

The first slice will be about 5.9 seconds and rest slices will be about 2.9 seconds. So the hls duration written in m3u8 file uses the maximum one.

> Also it's unclear why you want to convert to `Decimal` and then ceil, why not ceil immediately?

 I will update it later. TBH I'm just new to C# and have been relying on the experience of other languages ​​and google to do these ;)

Reply:
Don't worry too much about `m3u8` produced by ffmpeg, it's currently ignored anyway :(

Also, I strongly suggest changing it to what I suggested above - it would give you same 3 seconds but is "more correct" IMHO.